File: 2021/0118_generate/0818/generate.rs
impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        assert!(num_rows >= 0);
        gen_pascal_triangle()
            .take(num_rows as usize)
            .collect()
    }
}

fn gen_pascal_triangle() -> impl Iterator<Item = Vec<i32>> {
    std::iter::successors(Some(vec![1]), |row| {
        vec![1].into_iter()
            .chain(row.windows(2).map(|v| v.iter().sum()))
            .chain(vec![1].into_iter())
            .collect::<Vec<i32>>()
            .into()
    })
}

File: 2021/0118_generate/0818/generate.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        # 前两行都是1，不用修改, 从第三行开始: range(2, numRows)
        # 每一行首尾元素为1，不用处理: range(1, len(dp[i]) - 1)
        dp = [[1] * (i + 1) for i in range(numRows)]
        for i in range(2, numRows):
            for j in range(1, len(dp[i]) - 1):
                dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
        return dp

